### DISCUSSION TITLE
Draft order and checkout lines validation

### DISCUSSION THREAD

**Discussion by szymi-** [saleor/saleor#11827] [Q&A] [answered]
**Problem description**

It should be possible to validate draft order and checkout lines, basing on custom business requirements. 

Example business requirements:

- There is a product that can be added only once (only one piece can be bought in one order)
- There is a product that can be added only if another product of a specific type has already been added
- In certain circumstances, it should only be possible to only add a specific combination of products and / or product quantities

Apart from validation, it should be possible to edit draft order and checkout lines on the fly, if certain conditions are met. 

Example line edits on the fly: 

- If a certain product has been removed, some other product should also automatically be removed
- If a certain product has been added, some other product should also automatically be added
- In certain circumstances, some product should be added or removed automatically
- If only one piece of a product is allowed per order but more were added, extra pieces should be removed

**Proposed solution**

One way to achieve such validation could be to use an external app. For that, synchronous webhooks for `DRAFT_ORDER_LINES_UPDATED` and `CHECKOUT_LINES_UPDATED` events would be needed.

**Comment by timuric** [comment, marked as answer]
There is a danger of making checkout mutations slow by adding such webhook, and would create a lot of overhead in terms of API calls, as add to bag is a very popular action in storefronts.

Another issue is that It would not be a great user experience if customers would click "add to bag" and see an error, therefore you would want to add such business logic to the storefront anyway (which would be a duplicate code), so customers can see the proper message and UI. 

If you want to centralize such business logic across different applications, then writing a dedicated microservice would more efficient (faster response times), straightforward to document, debug and maintain.
